### PR TITLE
This change introduces "overlapping" indexes for locals debug info in mo...

### PR DESCRIPTION
**Pull request by carlokok**
This change introduces "overlapping" indexes for locals debug info in mono for the runtime. The debug info format and the soft debugger have logic already that makes this work, the only thing missing was to make mono return the index of locals with debug info instead of just of just one item by index.

This change should be backward/forward compatible (ie older debugger/newer debugger and same for runtime).

What this fixes:
Code like

```
void test() {
  {
    int a = 0;
    Console.WriteLine("Got " + a);
  }
  {
    int b = 0;
    Console.WriteLine("Got " + b);
  }
}
```

Where "a" and "b" reuse the same local variable slot.

Before this change "a" will show as a debug entry when it's in that scope. And in the scope for "b" nothing is shown at all. After this change both show as they should.

This change is released under the MIT license.

### CODE REVIEW

**Comment by monoadmin** [comment]
Hello! I'm the build bot for the Mono project. I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done. Contributors can ignore this message.

File: mono/mini/debugger-agent.c [hunk 1]
```diff
@@ -8485,41 +8485,78 @@ method_commands_internal (int command, MonoMethod *method, MonoDomain *domain, g
...
-					break;
-			if (j < num_locals && locals->locals [j].block) {
-				buffer_add_int (buf, locals->locals [j].block->start_offset);
-				buffer_add_int (buf, locals->locals [j].block->end_offset);
-			} else {
-				buffer_add_int (buf, 0);
-				buffer_add_int (buf, header->code_size);
-			}
-		}
+        if (!locals || !CHECK_PROTOCOL_VERSION (2, 41)) {
```
Comment (jbevain) [inline mono/mini/debugger-agent.c]: Tabs and spaces are mixed up apparently :)

**Comment by carlokok** [comment]
@jbevain: should be fixed

**Comment by migueldeicaza** [comment]
This really needs some tests showing the new behavior, to prevent regressions.

**Comment by carlokok** [comment]
Testcase added; the reason for a tiny binary.exe/mdb and not source is because mono doesn't reuse locals, nor can current cecil generate this properly on the fly, I hope this is oke.

**Comment by Therzok** [comment]
Heya!

Please don't add binaries to the repo. Integrate them in the build script as seen [here](https://github.com/mono/mono/blob/master/mcs/class/Mono.Debugger.Soft/Makefile#L16-L22).

**Comment by carlokok** [comment]
@Therzok: Problem is that mcs can't generate binaries like this. (Nor can cecil)

**Comment by alexrp** [comment]
Can the test case be expressed in IL?

**Comment by alexrp** [comment]
BTW, please fold 31369c1d8080bbff46347de04b607f8e3a32e228 into 1a80cf60901ef038179e29b27f9c755902300b83 for clean history.

**Comment by alexrp** [comment]
whitelist

**Comment by carlokok** [comment]
Oke, I've squashed the whitespace & original commit into one. As for the testcase: ILAsm doesn't support scopes yet. Upstream cecil does, but the version mono uses doesn't.

**Comment by alexrp** [comment]
Unfortunately, ILAsm still uses the ancient PEAPI library, so updating Cecil wouldn't do much good...

**Comment by carlokok** [comment]
Right; I mean I can create code that creates an exe like this using latest (upstream) cecil; but not the one used by mono, which is 7+ months behind.

**Comment by alexrp** [comment]
OK, I think adding binaries in this case is not the end of the world. They are highly unlikely to ever change and are fairly small. The problem with binaries is usually that they make the repo larger over time as they are updated.

What's everyone's thoughts on it given this?

In any case, please add a comment to `DebugLocals.cs` making it clear that this file can't be compiled with `mcs` (and maybe `csc` too? I don't know).

**Comment by Therzok** [comment]
3kb is not much, so binaries are ok in this case.

**Comment by tritao** [comment]
I also think binaries for test cases are perfectly fine, though if possible it's best to come up with a way to generate them from a text representation since that makes them easier to debug.

**Comment by alexrp** [comment]
@vargaz do you want to review the `debugger-agent.c` changes?

**Comment by alexrp** [comment]
This has since been fixed in master: https://github.com/mono/mono/commit/7efe163dcd2d0286e41c55924b95b477eba2a0b7

So, closing this.

**Comment by akoeplinger** [comment]
@alexrp what about the test cases in this PR? They seem valuable to me.

**Comment by carlokok** [comment]
Not sure how this solves it, the original issue was that you had 1 local variable slot, and two entries in the debug table (with different ranges). Before it wrote a item (the first match). Now it writes both but never the index:
```
buffer_add_int (buf, num_locals);

/* Types */
for (i = 0; i < num_locals; ++i) 
	buffer_add_typeid (buf, domain, mono_class_from_mono_type (header->locals [locals->locals [i].index]));
/* Names */
for (i = 0; i < num_locals; ++i)
	buffer_add_string (buf, locals->locals [i].name);
/* Scopes */
for (i = 0; i < num_locals; ++i) {
	if (locals->locals [i].block) {
		buffer_add_int (buf, locals->locals [i].block->start_offset);
		buffer_add_int (buf, locals->locals [i].block->end_offset);
	} else {
		buffer_add_int (buf, 0);
		buffer_add_int (buf, header->code_size);
	}
}
```

Unless I'm misunderstanding something, the SDB/debugger api has no way to find out which var goes in what slot? (my original patch changed both because of this)

**Comment by alexrp** [comment]
@carlokok you're right, it was me who misunderstood. (Sorry for the slow reply, I must've missed the email notification here.)

@vargaz any thoughts on this?

**Comment by vargaz** [comment]
What compiler generates such il code/debug info ? It shouldn't be done since it makes it harder for the JIT for example to generate good code since the merged variables have bigger live ranges, etc.

**Comment by carlokok** [comment]
My compiler (RemObjects Elements). But does that really matter? The debug info supports it but the runtime itself doesn't (and doesn't the runtime evaluate live ranges for locals in any case?)

**Comment by lewurm** [comment]
@carlokok can you rebase your changes please?

**Comment by lewurm** [comment]
hey @carlokok, sorry to ask again for it, but can you rebase this PR once again? After that, please ping me or @vargaz for a review.

**Comment by lewing** [comment]
This PR is inactive and has conflicts.  Feel free to reopen to restart review

**Pull request closed without merging**